### PR TITLE
add pixpro

### PR DESCRIPTION
**Pull request by lzzyzlbb**

### CODE REVIEW

File: passl/modeling/architectures/pixpro.py [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+
+from ...modules.init import init_backbone_weight
+from ...modules import freeze_batchnorm_statictis
+from .builder import MODELS
+from ..backbones import build_backbone
+from ..necks import build_neck
+from ..heads import build_head
+
+
+class Identity(nn.Layer):
```
Comment (LielinJiang) [inline passl/modeling/architectures/pixpro.py:27]: paddle有这个类，可以直接import使用

File: passl/modeling/architectures/pixpro.py [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+            self.backbone_k = nn.SyncBatchNorm.convert_sync_batchnorm(self.backbone_k)
+            self.neck = nn.SyncBatchNorm.convert_sync_batchnorm(self.neck)
+            self.neck_k = nn.SyncBatchNorm.convert_sync_batchnorm(self.neck_k)
+            
+        self.K = int(num_instances * 1. / gpu_num / batch_size * epochs)
+        self.k = int(num_instances * 1. / gpu_num / batch_size * (start_epoch - 1))
+        if self.pixpro_transform_layer == 0:
+            self.value_transform = Identity()
+        elif self.pixpro_transform_layer == 1:
+            self.value_transform = nn.Conv2D(256, 256, kernel_size=1, stride=1, padding=0, bias_attr=True, weight_attr=nn.initializer.KaimingUniform())
```
Comment (LielinJiang) [inline passl/modeling/architectures/pixpro.py:139]: 感觉单行太长，确认下是否使用了pre-commit

File: passl/datasets/imagenet.py [hunk 3]
```diff
@@ -79,6 +80,101 @@ def evaluate(self, preds, labels, topk=(1, 5)):
...
         return eval_res
 
+
+class PixProTransformCompose(object):
+    """Composes several transforms together.
+
+    Args:
+        transforms (list of ``Transform`` objects): list of transforms to compose.
+
+    Example:
```
Comment (LielinJiang) [inline passl/datasets/imagenet.py:90]: 改成paddle主框架的 example 风格吧

File: passl/datasets/preprocess/cv2_trans.py [hunk 4]
```diff
@@ -223,3 +226,120 @@ def decode_and_random_crop(self,img):
...
+class RandomResizedCropCoord(object):
+    """Crop the given PIL Image to random size and aspect ratio.
+
+    A crop of random size (default: of 0.08 to 1.0) of the original size and a random
+    aspect ratio (default: of 3/4 to 4/3) of the original aspect ratio is made. This crop
+    is finally resized to given size.
+    This is popularly used to train the Inception networks.
+
+    Args:
+        size: expected output size of each edge
```
Comment (LielinJiang) [inline passl/datasets/preprocess/cv2_trans.py:270]: 参数类型

File: passl/modeling/necks/base_neck.py [hunk 5]
```diff
@@ -255,6 +255,35 @@ def forward(self, x):
...
+
+    def forward(self, x):
+        x = self.linear1(x)
+        x = self.bn1(x)
+        x = self.relu1(x)
+        x = self.linear2(x)
+
+        return x
+
+
```
Comment (LielinJiang) [inline passl/modeling/necks/base_neck.py:289]: 这里的空行和之前的单行过长应该都是没有使用pre-commit

File: passl/modeling/architectures/pixpro.py (hunk 1 above)
Comment (lzzyzlbb) [reply passl/modeling/architectures/pixpro.py]: 已修改

File: passl/modeling/architectures/pixpro.py (hunk 2 above)
Comment (lzzyzlbb) [reply passl/modeling/architectures/pixpro.py]: 已修改

File: passl/datasets/imagenet.py (hunk 3 above)
Comment (lzzyzlbb) [reply passl/datasets/imagenet.py:89]: 已修改

File: passl/datasets/preprocess/cv2_trans.py (hunk 4 above)
Comment (lzzyzlbb) [reply passl/datasets/preprocess/cv2_trans.py:289]: 已修改

File: passl/modeling/necks/base_neck.py (hunk 5 above)
Comment (lzzyzlbb) [reply passl/modeling/necks/base_neck.py]: 已修改

**Review by LielinJiang — DISMISSED** [review]

**Review by LielinJiang — DISMISSED** [review]

File: passl/solver/optimizer.py [hunk 6]
```diff
@@ -23,3 +24,54 @@
...
+        _lars_weight_decay = self._lars_weight_decay
+        param_name = param_and_grad[0].name
+        if len(param_and_grad[0].shape) == 1:
+            _lars_weight_decay = 0.0
+
+        velocity_acc = self._get_accumulator(self._velocity_acc_str,
+                                             param_and_grad[0])
+        lr = self._create_param_lr(param_and_grad)
+
+        find_master = self._multi_precision and param_and_grad[
```
[the remark this answers is not in the document]
Comment (firestonelib) [reply passl/solver/optimizer.py:41]: 我验证了下你的代码，这里 shape==1 的只有 batch norm 和 conv2d_54.b0, conv2d_110.b0，其他的conv2d层没有 b0，所以可以参考 moco_byol 配置文件，使用 LarsMomentumOptimizer，设置 exclude_from_weight_decay: ['batch_norm', '.b_0']

File: configs/pixpro/pixpro_base_r50_100ep.yaml [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+  epochs: 100
+  start_epoch: 1
+  gpu_num: 8
+  pixpro_ins_loss_weight: 0.
+
+dataloader:
+  train:
+    num_workers: 4
+    sampler:
+      batch_size: 2
```
Comment (firestonelib) [inline configs/pixpro/pixpro_base_r50_100ep.yaml:37]: 按照paper设置，总batch-size1024， 单机8卡 batch_size 是不是应该等于 128 ？

**Review by firestonelib — APPROVED** [review]
LGTM
